### PR TITLE
Add additional type bounds to Impl traits

### PR DESCRIPTION
**Pull request by felinira**
Mirrors https://github.com/gtk-rs/gtk-rs-core/pull/1519
See https://github.com/gtk-rs/gtk-rs-core/issues/1515

Open questions:
- [x] `GInitiallyUnowned`: Technically a parent class, but we mostly ignore it in other places already
- [ ] Interfaces, especially the ones implemented by `GtkWidget` are currently not covered by the implementing classes. Should they? One normally doesn't need to reimplement them, but we need to check whether we are required to specify them for safety reasons anyway

### CODE REVIEW

**Comment by sdroege** [comment]
>     * [ ]  `GInitiallyUnowned`: Technically a parent class, but we mostly ignore it in other places already

I would ignore it here. It doesn't have its own impl trait, its own API, or anything. It's just a strange marker class :)

>     * [ ]  Interfaces, especially the ones implemented by `GtkWidget` are currently not covered by the implementing classes. Should they? One normally doesn't need to reimplement them, but we need to check whether we are required to specify them for safety reasons anyway

What do you mean, can you give a concrete example?

File: examples/virtual_methods/base_button/mod.rs [hunk 1]
```diff
@@ -60,7 +60,12 @@ impl<O: IsA<BaseButton>> BaseButtonExt for O {}
 /// implement.
 ///
 /// See `derived_button/imp.rs` for how to override virtual methods.
-pub trait BaseButtonImpl: ButtonImpl {
+pub trait BaseButtonImpl: ButtonImpl
+where
+    <Self as ObjectSubclass>::Type: IsA<glib::Object>,
+    <Self as ObjectSubclass>::Type: IsA<gtk::Widget>,
+    <Self as ObjectSubclass>::Type: IsA<gtk::Button>,
```
Comment (bilelmoussaoui) [inline examples/virtual_methods/base_button/mod.rs:67]: Doesn't this cover the two previous statements?

File: gtk4/src/subclass/accessible_range.rs [hunk 2]
```diff
@@ -6,20 +6,25 @@
 
 use glib::translate::*;
 
-use crate::{ffi, prelude::*, subclass::prelude::*, AccessibleRange};
+use crate::{ffi, prelude::*, subclass::prelude::*, Accessible, AccessibleRange};
 
-pub trait AccessibleRangeImpl: WidgetImpl {
+pub trait AccessibleRangeImpl: AccessibleImpl
```
[the remark this answers is not in the document]
Comment (bilelmoussaoui) [reply gtk4/src/subclass/accessible_range.rs]: But can you have accessible GObjects ?

File: gtk4/src/subclass/orientable.rs [hunk 3]
```diff
@@ -4,8 +4,18 @@
 //! Traits intended for implementing the [`Orientable`](crate::Orientable)
 //! interface.
 
-use crate::{subclass::prelude::*, Orientable};
+use crate::{prelude::*, subclass::prelude::*, Orientable};
 
-pub trait OrientableImpl: WidgetImpl {}
+pub trait OrientableImpl: ObjectImpl
```
[the remark this answers is not in the document]
Comment (bilelmoussaoui) [reply gtk4/src/subclass/orientable.rs]: same here

File: gtk4/src/subclass/accessible_range.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (bilelmoussaoui) [reply gtk4/src/subclass/accessible_range.rs]: Right, I am not sure where you can have an accessible object still. but let us go with this for now, to be consistent with AccessibleImpl. We can let someone else deal with the rest in the future if needed.

**Comment by bilelmoussaoui** [comment]
> We will now start requiring everyone to start adding all their parent classes here via @extends. But should we also require those interfaces to be specified as well? As in, add additional constraints to WidgetImpl to include IsA<Accessible>, IsA<Buildable> and IsA<ConstraintTarget>.

I think that is the conceptually correct thing to do but it will make every app developer's life more miserable given our poor subclassing code infrastructure.

On the other hand, enforcing things to be correctly handled would mean an easier path forward for some more automation using macros.


So for my side, I will just take the risk and do that. Could you give it a try in a separate commit to see the impact?

**Comment by sdroege** [comment]
I think requiring this is more correct and we should enforce it.

**Comment by sdroege** [comment]
@felinira This also requires some additional change in gtk-rs-core or not? If so, we can also do it in a separate PR if you prefer

**Comment by bilelmoussaoui** [comment]
> @felinira This also requires some additional change in gtk-rs-core or not? If so, we can also do it in a separate PR if you prefer

This cannot affect gtk-rs-core, there is no base object that implements some interfaces and have various sub-types. Maybe gstreamer-rs though :)

**Comment by sdroege** [comment]
Ah I thought one of the GIO interfaces maybe :)

**Comment by sdroege** [comment]
Let's get this in then and do the interfaces separately or what do you think?

**Comment by bilelmoussaoui** [comment]
Sure

**Comment by bilelmoussaoui** [comment]
Thanks @felinira

**Pull request merged**